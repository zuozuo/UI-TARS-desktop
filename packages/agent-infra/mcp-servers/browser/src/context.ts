/**
 * Copyright (c) 2025 Bytedance, Inc. and its affiliates.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Page } from 'puppeteer-core';
import { store } from './store.js';
import { ResourceContext, ToolContext } from './typings.js';
import { ensureBrowser } from './utils/browser.js';
import {
  getBuildDomTreeScript,
  parseNode,
  type RawDomTreeNode,
  DOMElementNode,
  createSelectorMap,
} from '@agent-infra/browser-use';

export class Context {
  async getResourceContext(): Promise<ResourceContext> {
    return {
      logger: store.logger,
    };
  }

  async getToolContext(): Promise<ToolContext | null> {
    const { logger, globalConfig } = store;

    const initialBrowser = await ensureBrowser();
    const { browser, currTabsIdx } = initialBrowser;
    let { page } = initialBrowser;

    page.removeAllListeners('popup');
    page.on('popup', async (popup) => {
      if (popup) {
        logger.info(`popup page: ${popup.url()}`);
        await popup.bringToFront();
        page = popup;
        store.globalPage = popup;
      }
    });

    return {
      page,
      currTabsIdx,
      browser,
      logger,
      contextOptions: globalConfig.contextOptions || {},
      buildDomTree: this.buildDomTree,
    };
  }

  async buildDomTree(page: Page) {
    const logger = store.logger;

    try {
      // check if the buildDomTree script is already injected
      const existBuildDomTreeScript = await page.evaluate(
        /* istanbul ignore next */ () => {
          return typeof window.buildDomTree === 'function';
        },
      );
      if (!existBuildDomTreeScript) {
        const injectScriptContent = getBuildDomTreeScript();
        await page.evaluate(
          /* istanbul ignore next */ (script) => {
            const scriptElement = document.createElement('script');
            scriptElement.textContent = script;
            document.head.appendChild(scriptElement);
          },
          injectScriptContent,
        );
      }

      const rawDomTree = await page.evaluate(
        /* istanbul ignore next */ () => {
          // Access buildDomTree from the window context of the target page
          return window.buildDomTree({
            doHighlightElements: true,
            focusHighlightIndex: -1,
            viewportExpansion: 0,
          });
        },
      );
      if (rawDomTree !== null) {
        const elementTree = parseNode(rawDomTree as RawDomTreeNode);
        if (elementTree !== null && elementTree instanceof DOMElementNode) {
          const clickableElements = elementTree.clickableElementsToString();
          store.selectorMap = createSelectorMap(elementTree);

          return {
            clickableElements,
            elementTree,
            selectorMap: store.selectorMap,
          };
        }
      }
      return null;
    } catch (error) {
      logger.error('Error building DOM tree:', error);
      return null;
    }
  }
}
