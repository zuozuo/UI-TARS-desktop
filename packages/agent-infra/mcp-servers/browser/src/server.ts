/**
 * The following code is modified based on
 * https://github.com/modelcontextprotocol/servers/blob/main/src/puppeteer/index.ts
 *
 * MIT License
 * Copyright (c) 2024 Anthropic, PBC
 * https://github.com/modelcontextprotocol/servers/blob/main/LICENSE
 */
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import {
  CallToolResult,
  ImageContent,
  TextContent,
} from '@modelcontextprotocol/sdk/types.js';

import { Logger, BaseLogger } from '@agent-infra/logger';
import { z } from 'zod';
import { Page } from '@agent-infra/browser';
import { removeHighlights, locateElement } from '@agent-infra/browser-use';
import merge from 'lodash.merge';
import { defineTools, delay, getDownloadSuggestion } from './utils/utils.js';
import { Browser, ElementHandle, KeyInput } from 'puppeteer-core';
import { keyInputValues } from './constants.js';
import {
  GlobalConfig,
  ResourceContext,
  ToolContext,
  ToolDefinition,
} from './typings.js';
import {
  screenshots,
  getScreenshots,
  registerResources,
} from './resources/index.js';
import { store } from './store.js';
import { Context } from './context.js';

// tools
import visionTools from './tools/vision.js';
import downloadTools from './tools/download.js';
import navigateTools from './tools/navigate.js';
import contentTools from './tools/content.js';
import tabsTools from './tools/tabs.js';

function setConfig(config: GlobalConfig = {}) {
  store.globalConfig = merge({}, store.globalConfig, config);
  if (config.logger) {
    store.logger = config.logger as Logger;
  }
}
function getConfig(): GlobalConfig {
  return store.globalConfig;
}

async function setInitialBrowser(
  _browser?: Browser,
  _page?: Page,
): Promise<{ browser: Browser; page: Page }> {
  const { logger } = store;

  // priority 1: use provided browser and page
  if (_browser) {
    logger.info('Using global browser');
    store.globalBrowser = _browser;
  }
  if (_page) {
    store.globalPage = _page;
  }

  return {
    browser: store.globalBrowser!,
    page: store.globalPage!,
  };
}

export const getBrowser = () => {
  return { browser: store.globalBrowser, page: store.globalPage };
};

export const toolsMap = defineTools({
  browser_screenshot: {
    name: 'browser_screenshot',
    description: 'Take a screenshot of the current page or a specific element',
    inputSchema: z.object({
      name: z.string().optional().describe('Name for the screenshot'),
      selector: z
        .string()
        .optional()
        .describe('CSS selector for element to screenshot'),
      index: z
        .number()
        .optional()
        .describe('index of the element to screenshot'),
      width: z
        .number()
        .optional()
        .describe('Width in pixels (default: viewport width)'),
      height: z
        .number()
        .optional()
        .describe('Height in pixels (default: viewport height)'),
      fullPage: z
        .boolean()
        .optional()
        .describe('Full page screenshot (default: false)'),
      highlight: z
        .boolean()
        .optional()
        .default(false)
        .describe('Highlight the element'),
    }),
  },
  browser_click: {
    name: 'browser_click',
    description:
      'Click an element on the page, before using the tool, use `browser_get_clickable_elements` to get the index of the element, but not call `browser_get_clickable_elements` multiple times',
    inputSchema: z.object({
      // selector: z
      //   .string()
      //   .optional()
      //   .describe('CSS selector for element to click'),
      index: z.number().optional().describe('Index of the element to click'),
    }),
  },
  browser_form_input_fill: {
    name: 'browser_form_input_fill',
    description:
      "Fill out an input field, before using the tool, Either 'index' or 'selector' must be provided",
    inputSchema: z.object({
      selector: z.string().optional().describe('CSS selector for input field'),
      index: z.number().optional().describe('Index of the element to fill'),
      value: z.string().describe('Value to fill'),
    }),
  },
  browser_select: {
    name: 'browser_select',
    description:
      "Select an element on the page with index, Either 'index' or 'selector' must be provided",
    inputSchema: z.object({
      index: z.number().optional().describe('Index of the element to select'),
      selector: z
        .string()
        .optional()
        .describe('CSS selector for element to select'),
      value: z.string().describe('Value to select'),
    }),
  },
  browser_hover: {
    name: 'browser_hover',
    description:
      "Hover an element on the page, Either 'index' or 'selector' must be provided",
    inputSchema: z.object({
      index: z.number().optional().describe('Index of the element to hover'),
      selector: z
        .string()
        .optional()
        .describe('CSS selector for element to hover'),
    }),
  },
  browser_evaluate: {
    name: 'browser_evaluate',
    description: 'Execute JavaScript in the browser console',
    inputSchema: z.object({
      script: z.string().describe('JavaScript code to execute'),
    }),
  },
  browser_get_clickable_elements: {
    name: 'browser_get_clickable_elements',
    description:
      "Get the clickable or hoverable or selectable elements on the current page, don't call this tool multiple times",
  },
  browser_scroll: {
    name: 'browser_scroll',
    description: 'Scroll the page',
    inputSchema: z.object({
      amount: z
        .number()
        .optional()
        .describe(
          'Pixels to scroll (positive for down, negative for up), if the amount is not provided, scroll to the bottom of the page',
        ),
    }),
  },
  browser_close: {
    name: 'browser_close',
    description:
      'Close the browser when the task is done and the browser is not needed anymore',
  },
  browser_press_key: {
    name: 'browser_press_key',
    description: 'Press a key on the keyboard',
    inputSchema: z.object({
      key: z
        .enum(keyInputValues as [string, ...string[]])
        .describe(
          `Name of the key to press or a character to generate, such as ${keyInputValues.join(
            ', ',
          )}`,
        ),
    }),
  },
});

type ToolNames = keyof typeof toolsMap;
type ToolInputMap = {
  [K in ToolNames]: (typeof toolsMap)[K] extends {
    inputSchema: infer S;
  }
    ? S extends z.ZodType<any, any, any>
      ? z.infer<S>
      : unknown
    : unknown;
};

const handleToolCall = async (
  ctx: Context,
  {
    name,
    arguments: toolArgs,
  }: {
    name: string;
    arguments: ToolInputMap[keyof ToolInputMap];
  },
): Promise<CallToolResult> => {
  const toolCtx = await ctx.getToolContext();

  if (!toolCtx?.page) {
    return {
      content: [{ type: 'text', text: 'Page not found' }],
      isError: true,
    };
  }

  const { page, browser, logger } = toolCtx;

  const handlers: {
    [K in ToolNames]: (args: ToolInputMap[K]) => Promise<CallToolResult>;
  } = {
    browser_screenshot: async (args) => {
      // if highlight is true, build the dom tree with highlights
      try {
        if (args.highlight) {
          await ctx.buildDomTree(page);
        } else {
          await removeHighlights(page);
        }
      } catch (error) {
        logger.warn('[browser_screenshot] Error building DOM tree:', error);
      }

      const width = args.width ?? page.viewport()?.width ?? 800;
      const height = args.height ?? page.viewport()?.height ?? 600;
      await page.setViewport({ width, height });

      let screenshot: string | undefined;
      if (args.selector) {
        screenshot = await (args.selector
          ? (await page.$(args.selector))?.screenshot({ encoding: 'base64' })
          : undefined);
      } else if (args.index !== undefined) {
        const elementNode = store.selectorMap?.get(Number(args?.index));
        const element = await locateElement(page, elementNode!);

        screenshot = await (element
          ? element.screenshot({ encoding: 'base64' })
          : undefined);
      }

      // if screenshot is still undefined, take a screenshot of the whole page
      screenshot =
        screenshot ||
        (await page.screenshot({
          encoding: 'base64',
          fullPage: args.fullPage ?? false,
        }));

      // if screenshot is still undefined, return an error
      if (!screenshot) {
        return {
          content: [
            {
              type: 'text',
              text: `Element not found: ${args.selector || args.index}`,
            },
          ],
          isError: true,
        };
      }

      const name = args?.name ?? 'undefined';

      screenshots.set(name, screenshot as string);

      const dimensions = args.fullPage
        ? await page.evaluate(
            /* istanbul ignore next */ () => ({
              width: Math.max(
                document.documentElement.scrollWidth,
                document.documentElement.clientWidth,
                document.body.scrollWidth,
              ),
              height: Math.max(
                document.documentElement.scrollHeight,
                document.documentElement.clientHeight,
                document.body.scrollHeight,
              ),
            }),
          )
        : { width, height };

      return {
        content: [
          {
            type: 'text',
            text: args.fullPage
              ? `Screenshot of the whole page taken at ${dimensions.width}x${dimensions.height}`
              : `Screenshot '${name}' taken at ${dimensions.width}x${dimensions.height}`,
          } as TextContent,
          {
            type: 'image',
            data: screenshot,
            mimeType: 'image/png',
          } as ImageContent,
        ],
        isError: false,
      };
    },
    browser_get_clickable_elements: async (args) => {
      if (!page) {
        return {
          content: [{ type: 'text', text: 'Page not found' }],
          isError: true,
        };
      }

      try {
        const { clickableElements } = (await ctx.buildDomTree(page)) || {};
        await removeHighlights(page);
        if (clickableElements) {
          return {
            content: [
              {
                type: 'text',
                text: clickableElements,
              },
            ],
            isError: false,
          };
        }
        return {
          content: [
            {
              type: 'text',
              text: 'Failed to parse DOM tree',
            },
          ],
          isError: false,
        };
      } catch (error) {
        return {
          content: [{ type: 'text', text: (error as Error).message }],
          isError: true,
        };
      }
    },
    browser_click: async (args) => {
      try {
        const downloadsBefore = store.downloadedFiles.length;

        let element: ElementHandle<Element> | null = null;
        if (args?.index !== undefined) {
          const elementNode = store.selectorMap?.get(Number(args?.index));
          if (elementNode?.highlightIndex !== undefined) {
            await removeHighlights(page);
          }

          element = await locateElement(page, elementNode!);
        }
        // else if (args.selector) {
        //   element = await page.$(args.selector);
        //   // locateElement
        //   await scrollIntoViewIfNeeded(element!);
        // }
        else {
          return {
            content: [
              {
                type: 'text',
                text: `Element index ${args?.index} not found`,
              },
            ],
            isError: true,
          };
        }

        try {
          await Promise.race([
            element?.click(),
            new Promise((_, reject) =>
              setTimeout(() => reject(new Error('Click timeout')), 5000),
            ),
          ]);

          await delay(200);

          const currentDownloadSuggestion = getDownloadSuggestion(
            downloadsBefore,
            store.downloadedFiles,
            store.globalConfig.outputDir!,
          );

          return {
            content: [
              {
                type: 'text',
                text: `Clicked element: ${args.index}${currentDownloadSuggestion}`,
              },
            ],
            isError: false,
          };
        } catch (error) {
          // Second attempt: Use evaluate to perform a direct click
          logger.error('Failed to click element, trying again', error);
          try {
            await element?.evaluate((el) => (el as HTMLElement).click());

            await delay(200);

            const currentDownloadSuggestion = getDownloadSuggestion(
              downloadsBefore,
              store.downloadedFiles,
              store.globalConfig.outputDir!,
            );

            return {
              content: [
                {
                  type: 'text',
                  text: `Clicked element: ${args.index}${currentDownloadSuggestion}`,
                },
              ],
              isError: false,
            };
          } catch (secondError) {
            return {
              content: [
                {
                  type: 'text',
                  text: `Failed to click element: ${secondError instanceof Error ? secondError.message : String(secondError)}`,
                },
              ],
              isError: true,
            };
          }
        }
      } catch (error) {
        return {
          isError: true,
          content: [
            {
              type: 'text',
              text: `Failed to click element: ${args.index}. Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
        };
      }
    },
    browser_form_input_fill: async (args) => {
      try {
        if (args.index !== undefined) {
          const elementNode = store.selectorMap?.get(Number(args?.index));

          if (elementNode?.highlightIndex !== undefined) {
            await removeHighlights(page);
          }

          const element = await locateElement(page, elementNode!);

          if (!element) {
            return {
              content: [{ type: 'text', text: 'No form input found' }],
              isError: true,
            };
          }
          await element?.type(args.value);
        } else if (args.selector) {
          await page.waitForSelector(args.selector);
          await page.type(args.selector, args.value);
        } else {
          return {
            content: [
              {
                type: 'text',
                text: 'Either selector or index must be provided',
              },
            ],
            isError: true,
          };
        }

        return {
          content: [
            {
              type: 'text',
              text: `Filled ${args.selector ? args.selector : args.index} with: ${args.value}`,
            },
          ],
          isError: false,
        };
      } catch (error) {
        return {
          content: [
            {
              type: 'text',
              text: `Failed to fill ${args.selector ? args.selector : args.index}: ${(error as Error).message}`,
            },
          ],
          isError: true,
        };
      }
    },
    browser_select: async (args) => {
      try {
        if (args.index !== undefined) {
          const elementNode = store.selectorMap?.get(Number(args?.index));

          if (elementNode?.highlightIndex !== undefined) {
            await removeHighlights(page);
          }

          const element = await locateElement(page, elementNode!);

          if (!element) {
            return {
              content: [{ type: 'text', text: 'No form input found' }],
              isError: true,
            };
          }

          await element?.select(args.value);
        } else if (args.selector) {
          await page.waitForSelector(args.selector);
          await page.select(args.selector, args.value);
        } else {
          return {
            content: [
              {
                type: 'text',
                text: `No selector ${args.selector} or index ${args.index} provided`,
              },
            ],
            isError: true,
          };
        }

        return {
          content: [
            {
              type: 'text',
              text: `Selected ${args.selector ? args.selector : args.index} with: ${args.value}`,
            },
          ],
          isError: false,
        };
      } catch (error) {
        return {
          content: [
            {
              type: 'text',
              text: `Failed to select ${args.selector ? args.selector : args.index}: ${(error as Error).message}`,
            },
          ],
          isError: true,
        };
      }
    },
    browser_hover: async (args) => {
      try {
        if (args.index !== undefined) {
          const elementNode = store.selectorMap?.get(Number(args?.index));

          if (elementNode?.highlightIndex !== undefined) {
            await removeHighlights(page);
          }

          const element = await locateElement(page, elementNode!);

          if (!element) {
            return {
              content: [{ type: 'text', text: 'No element found' }],
              isError: true,
            };
          }
          await element?.hover();
        } else if (args.selector) {
          await page.waitForSelector(args.selector);
          await page.hover(args.selector);
        } else {
          return {
            content: [
              {
                type: 'text',
                text: `No selector ${args.selector} or index ${args.index} provided`,
              },
            ],
            isError: true,
          };
        }

        return {
          content: [
            {
              type: 'text',
              text: `Hovered ${args.selector ? args.selector : args.index}`,
            },
          ],
          isError: false,
        };
      } catch (error) {
        return {
          content: [
            {
              type: 'text',
              text: `Failed to hover ${args.selector ? args.selector : args.index}: ${(error as Error).message}`,
            },
          ],
          isError: true,
        };
      }
    },
    browser_evaluate: async (args) => {
      try {
        await page.evaluate(
          /* istanbul ignore next */ () => {
            window.mcpHelper = {
              logs: [],
              originalConsole: { ...console },
            };

            ['log', 'info', 'warn', 'error'].forEach((method) => {
              (console as any)[method] = (...args: any[]) => {
                window.mcpHelper.logs.push(`[${method}] ${args.join(' ')}`);
                (window.mcpHelper.originalConsole as any)[method](...args);
              };
            });
          },
        );
        /* istanbul ignore next */
        const result = await page.evaluate(args.script);

        const logs = await page.evaluate(
          /* istanbul ignore next */ () => {
            Object.assign(console, window.mcpHelper.originalConsole);
            const logs = window.mcpHelper.logs;
            delete (window as any).mcpHelper;
            return logs;
          },
        );

        return {
          content: [
            {
              type: 'text',
              text: `Execution result:\n${JSON.stringify(result, null, 2)}\n`,
            },
          ],
          isError: false,
        };
      } catch (error) {
        return {
          content: [
            {
              type: 'text',
              text: `Script execution failed: ${(error as Error).message}`,
            },
          ],
          isError: true,
        };
      }
    },
    browser_scroll: async (args) => {
      try {
        const scrollResult = await page.evaluate(
          /* istanbul ignore next */ (amount) => {
            const beforeScrollY = window.scrollY;
            if (amount) {
              window.scrollBy(0, amount);
            } else {
              window.scrollBy(0, window.innerHeight);
            }

            // check if the page is scrolled the expected distance
            const actualScroll = window.scrollY - beforeScrollY;

            // check if the page is at the bottom
            const scrollHeight = Math.max(
              document.documentElement.scrollHeight,
              document.body.scrollHeight,
            );
            const scrollTop = window.scrollY;
            const clientHeight =
              window.innerHeight || document.documentElement.clientHeight;
            const isAtBottom =
              Math.abs(scrollHeight - scrollTop - clientHeight) <= 1;

            return {
              actualScroll,
              isAtBottom,
            };
          },
          args.amount,
        );

        return {
          content: [
            {
              type: 'text',
              text: `Scrolled ${scrollResult.actualScroll} pixels. ${
                scrollResult.isAtBottom
                  ? 'Reached the bottom of the page.'
                  : 'Did not reach the bottom of the page.'
              }`,
            },
          ],
          isError: false,
        };
      } catch (error) {
        return {
          content: [
            {
              type: 'text',
              text: `Failed to scroll: ${args.amount}`,
            },
          ],
          isError: true,
        };
      }
    },
    browser_close: async (args) => {
      try {
        await browser?.close();

        store.globalBrowser = null;
        store.globalPage = null;
        store.downloadedFiles = [];
        store.initialBrowserSetDownloadBehavior = false;

        return {
          content: [{ type: 'text', text: 'Closed browser' }],
          isError: false,
        };
      } catch (error) {
        return {
          content: [
            {
              type: 'text',
              text: `Failed to close browser: ${(error as Error).message}`,
            },
          ],
          isError: true,
        };
      }
    },
    browser_press_key: async (args) => {
      try {
        await page.keyboard.press(args.key as KeyInput);
        return {
          content: [{ type: 'text', text: `Pressed key: ${args.key}` }],
          isError: false,
        };
      } catch (error) {
        return {
          content: [{ type: 'text', text: `Failed to press key: ${args.key}` }],
          isError: true,
        };
      }
    },
  };

  if (handlers[name as ToolNames]) {
    return handlers[name as ToolNames](toolArgs as any);
  }

  return {
    content: [
      {
        type: 'text',
        text: `Unknown tool: ${name}`,
      },
    ],
    isError: true,
  };
};

function createServer(config: GlobalConfig = {}): McpServer {
  setConfig(config);

  const server = new McpServer(
    {
      name: 'Web Browser',
      version: process.env.VERSION || '0.0.1',
    },
    {
      capabilities: {
        logging: {},
      },
    },
  );

  // === Tools ===
  // Old
  const mergedToolsMap: Record<string, ToolDefinition> = {
    ...toolsMap,
  };

  const ctx = new Context();

  // New Tools
  const newTools = [
    ...navigateTools,
    ...contentTools,
    ...tabsTools,
    ...(config.vision ? visionTools : []),
    ...downloadTools,
  ];
  newTools.forEach((tool) => {
    server.registerTool(tool.name, tool.config as any, async (args) => {
      if (tool.skipToolContext) {
        return tool.handle(null, args);
      }

      const toolCtx = await ctx.getToolContext();

      if (!toolCtx?.page) {
        return {
          content: [{ type: 'text', text: 'Page not found' }],
          isError: true,
        };
      }

      // @ts-expect-error
      return tool.handle(toolCtx as ToolContext, args);
    });
  });

  // === Old Tools ===
  Object.entries(mergedToolsMap).forEach(([name, tool]) => {
    // @ts-ignore
    if (tool?.inputSchema) {
      server.tool(
        name,
        tool.description,
        // @ts-ignore
        tool.inputSchema?.innerType
          ? // @ts-ignore
            tool.inputSchema.innerType().shape
          : // @ts-ignore
            tool.inputSchema?.shape,
        // @ts-ignore
        async (args) => await handleToolCall(ctx, { name, arguments: args }),
      );
    } else {
      server.tool(
        name,
        tool.description,
        async (args) => await handleToolCall(ctx, { name, arguments: args }),
      );
    }
  });

  const resourceCtx: ResourceContext = {
    logger: store.logger,
  };

  // === Resources ===
  registerResources(server, resourceCtx);

  return server;
}

export {
  BaseLogger,
  createServer,
  getScreenshots,
  setConfig,
  getConfig,
  type GlobalConfig,
  setInitialBrowser,
};
