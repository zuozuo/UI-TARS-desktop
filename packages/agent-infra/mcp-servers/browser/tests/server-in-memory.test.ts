/*
 * Copyright (c) 2025 Bytedance, Inc. and its affiliates.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Jimp } from 'jimp';
import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  test,
} from 'vitest';
import express from 'express';
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { InMemoryTransport } from '@modelcontextprotocol/sdk/inMemory.js';
import { AddressInfo } from 'net';
import { createServer, type GlobalConfig } from '../src/server';

describe('MCP Server in memory', () => {
  test('listTools should return a list of tools', async () => {
    const client = new Client(
      {
        name: 'test client',
        version: '1.0',
      },
      {
        capabilities: {
          roots: {
            listChanged: true,
          },
        },
      },
    );

    const server = createServer({
      launchOptions: {
        headless: true,
      },
    } as GlobalConfig);
    const [clientTransport, serverTransport] =
      InMemoryTransport.createLinkedPair();

    await Promise.all([
      client.connect(clientTransport),
      server.connect(serverTransport),
    ]);

    const result = await client.listTools();

    expect(result.tools.map((tool) => tool.name).sort()).toMatchSnapshot();
  });

  test('listTools should return a list of tools with --vision', async () => {
    const client = new Client(
      {
        name: 'test client',
        version: '1.0',
      },
      {
        capabilities: {
          roots: {
            listChanged: true,
          },
        },
      },
    );

    const server = createServer({
      launchOptions: {
        headless: true,
      },
      vision: true,
    } as GlobalConfig);
    const [clientTransport, serverTransport] =
      InMemoryTransport.createLinkedPair();

    await Promise.all([
      client.connect(clientTransport),
      server.connect(serverTransport),
    ]);

    const result = await client.listTools();

    expect(result.tools.map((tool) => tool.name).sort()).toMatchSnapshot();
  });

  describe('call tools', () => {
    let client: Client;
    let app: express.Express;
    let httpServer: any;
    let baseUrl: string;

    beforeAll(async () => {
      app = express();

      // 添加测试页面路由
      app.get('/', (req, res) => {
        res.send(`
          <!DOCTYPE html>
          <html>
            <head><title>Test Page</title></head>
            <body>
              <h1>Test Page</h1>
              <input type="text" id="testInput" />
              <button id="testButton">Click Me</button>
              <a href="/page2">Go to Page 2</a>
              <select id="testSelect">
                <option value="1">Option 1</option>
                <option value="2">Option 2</option>
              </select>
            </body>
          </html>
        `);
      });

      app.get('/page2', (req, res) => {
        res.send(`
          <!DOCTYPE html>
          <html>
            <head><title>Page 2</title></head>
            <body>
              <h1>Page 2</h1>
              <a href="/">Back to Home</a>
            </body>
          </html>
        `);
      });

      httpServer = app.listen(0); // 使用随机可用端口
      const address = httpServer.address() as AddressInfo;
      baseUrl = `http://localhost:${address.port}`;
    });

    afterAll(async () => {
      await httpServer.close();
    });

    beforeEach(async () => {
      client = new Client(
        {
          name: 'test client',
          version: '1.0',
        },
        {
          capabilities: {
            roots: {
              listChanged: true,
            },
          },
        },
      );

      const server = createServer({
        launchOptions: {
          headless: true,
        },
      } as GlobalConfig);
      const [clientTransport, serverTransport] =
        InMemoryTransport.createLinkedPair();

      await Promise.all([
        client.connect(clientTransport),
        server.connect(serverTransport),
      ]);
    });

    afterEach(async () => {
      await client.callTool({
        name: 'browser_close',
        arguments: {},
      });
      await client.close();
    });

    test('browser_form_input_fill', async () => {
      await client.callTool({
        name: 'browser_navigate',
        arguments: {
          url: baseUrl,
        },
      });

      const result = await client.callTool({
        name: 'browser_form_input_fill',
        arguments: {
          value: 'input_value',
        },
      });
      expect(result.isError).toEqual(true);

      const resultSuccess = await client.callTool({
        name: 'browser_form_input_fill',
        arguments: {
          selector: 'input',
          value: 'input_value',
        },
      });

      expect(resultSuccess.isError).toEqual(false);
    });

    test('no vision tools', async () => {
      await expect(
        client.callTool({
          name: 'browser_vision_screen_click',
          arguments: {
            x: 100,
            y: 100,
            factors: [1000, 1000],
          },
        }),
      ).rejects.toThrowError(/not found/);

      await expect(
        client.callTool({
          name: 'browser_vision_screen_capture',
          arguments: {},
        }),
      ).rejects.toThrowError(/not found/);
    });

    test('set viewport size', async () => {
      const newClient = new Client(
        {
          name: 'test client',
          version: '1.0',
        },
        {
          capabilities: {
            roots: {
              listChanged: true,
            },
          },
        },
      );

      const defaultViewport = {
        width: 1280,
        height: 960,
      };

      const server = createServer({
        launchOptions: {
          headless: true,
          defaultViewport,
        },
        vision: true,
      } as GlobalConfig);
      const [clientTransport, serverTransport] =
        InMemoryTransport.createLinkedPair();

      await Promise.all([
        newClient.connect(clientTransport),
        server.connect(serverTransport),
      ]);

      await newClient.callTool({
        name: 'browser_navigate',
        arguments: {
          url: baseUrl,
        },
      });

      const results = await newClient.callTool({
        name: 'browser_screenshot',
        arguments: {
          name: 'test_screenshot',
        },
      });

      const { width, height } = await Jimp.fromBuffer(
        Buffer.from(results.content?.[1].data, 'base64'),
      );

      expect({
        width,
        height,
      }).toEqual(defaultViewport);

      const visionResults = await newClient.callTool({
        name: 'browser_vision_screen_capture',
        arguments: {},
      });

      const { width: visionWidth, height: visionHeight } =
        await Jimp.fromBuffer(
          Buffer.from(visionResults.content?.[1].data, 'base64'),
        );

      expect({
        width: visionWidth,
        height: visionHeight,
      }).toEqual(defaultViewport);
    });
  });

  describe('call vision tools', () => {
    let client: Client;
    beforeEach(async () => {
      client = new Client(
        {
          name: 'test client',
          version: '1.0',
        },
        {
          capabilities: {
            roots: {
              listChanged: true,
            },
          },
        },
      );

      const server = createServer({
        launchOptions: {
          headless: true,
        },
        vision: true,
      } as GlobalConfig);
      const [clientTransport, serverTransport] =
        InMemoryTransport.createLinkedPair();

      await Promise.all([
        client.connect(clientTransport),
        server.connect(serverTransport),
      ]);
    });

    afterEach(async () => {
      await client.close();
    });

    test('browser_vision_screen_click', async () => {
      const result = await client.callTool({
        name: 'browser_vision_screen_click',
        arguments: {
          x: 100,
          y: 100,
          factors: [1000, 1000],
        },
      });

      expect(result).toEqual({
        content: [
          {
            text: 'Vision click at 100, 100',
            type: 'text',
          },
        ],
        isError: false,
      });
    });
  });
});
