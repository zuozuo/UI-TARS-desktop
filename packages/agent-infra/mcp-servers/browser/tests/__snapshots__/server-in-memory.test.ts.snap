// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MCP Server in memory > listTools should return a list of tools 1`] = `
[
  "browser_click",
  "browser_close",
  "browser_close_tab",
  "browser_evaluate",
  "browser_form_input_fill",
  "browser_get_clickable_elements",
  "browser_get_download_list",
  "browser_get_markdown",
  "browser_get_text",
  "browser_go_back",
  "browser_go_forward",
  "browser_hover",
  "browser_navigate",
  "browser_new_tab",
  "browser_press_key",
  "browser_read_links",
  "browser_screenshot",
  "browser_scroll",
  "browser_select",
  "browser_switch_tab",
  "browser_tab_list",
]
`;

exports[`MCP Server in memory > listTools should return a list of tools with --vision 1`] = `
[
  "browser_click",
  "browser_close",
  "browser_close_tab",
  "browser_evaluate",
  "browser_form_input_fill",
  "browser_get_clickable_elements",
  "browser_get_download_list",
  "browser_get_markdown",
  "browser_get_text",
  "browser_go_back",
  "browser_go_forward",
  "browser_hover",
  "browser_navigate",
  "browser_new_tab",
  "browser_press_key",
  "browser_read_links",
  "browser_screenshot",
  "browser_scroll",
  "browser_select",
  "browser_switch_tab",
  "browser_tab_list",
  "browser_vision_screen_capture",
  "browser_vision_screen_click",
]
`;
